contracts/UserAuthority.sol

```diff
@@ -0,0 +1,233 @@
+pragma solidity <0.6.0;
+
+import "./interfaces/SignaturitUserInterface.sol";
+import "./libraries/Utils.sol";
+
+/*
+Gas to deploy: 2.896.394
+*/
+
+/*
+User's role:
+0 - User
+1 - Admin
+2 - Root
+
+- Everyone that create an user is also manager for that user
+- Root is manager of all the users
+
+Admins can:
+- create users
+- create admins
+- set role, reputation and validity for managed users
+
+Admins can't:
+- set role, validity and reputation of root
+- create root users
+- set role, reputation and validity of not managed users
+
+Root can:
+- create admins
+- create users
+- create root users
+- set role, reputation and validity of everyone
+- revoke management of users to admins
+
+User's reputation:
+Scale from 0 to 5, by default it sets to 3 to reflect neutrality
+
+- For each new user with contractAddress not null the rootAddress is set as manager on the user smart contract
+*/
+
+contract UserAuthority {
+    string constant private MANAGERS_KEY = "admitted-managers";
+
+    address public rootAddress;
+
+    struct User {
+        address contractAddress;
+        uint role;
+        uint reputation;
+        bool validity;
+        address[] managedUsers;
+    }
+    
+    mapping(address => User) users;
+
+    constructor () public {
+        rootAddress = msg.sender;
+
+        address[] memory tmpAddresses;
+
+        users[rootAddress] = User(address(0), 2, 5, true, tmpAddresses);
+    }
+
+    modifier allowedAdminCreation() {
+        require(
+            users[msg.sender].role > 1,
+            "The account can't perform this action"
+        );
+
+        _;
+    }
+
+    modifier allowedUserCreation() {
+        require(
+            users[msg.sender].role > 0,
+            "The account can't perform this action"
+        );
+
+        _;
+    }
+
+    modifier allowedEdit(address userAdr) {
+        require(
+            users[msg.sender].role >= users[userAdr].role &&
+            _isUserManager(userAdr, msg.sender),
+            "Unauthorized action"
+        );
+
+        _;
+    }
+
+    function createAdmin(
+        address adminAdr
+    )
+        public
+        allowedAdminCreation
+    {
+        address[] memory tmpAddresses;
+
+        users[adminAdr] = User(address(0), 1, 3, true, tmpAddresses);
+
+        users[msg.sender].managedUsers.push(adminAdr);
+    }
+
+    function createUser(
+        address userAdr
+    )
+        public
+        allowedUserCreation
+    {
+        address[] memory tmpAddresses;
+
+        users[userAdr] = User(address(0), 0, 3, true, tmpAddresses);
+
+        users[msg.sender].managedUsers.push(userAdr);
+    }
+
+    function setUserContract(
```

Comment:
I don't think will be a good idea merge this contract with the user, the responsability  of both contracts are different